Avoid removing plan with active marketplace subscriptions

gcsecsey opened the pull request:
#### Changes proposed in this Pull Request

* Added a new modal for removing plans that have active Marketplace subscriptions
* The modal has a "Remove Plan & All Subscriptions" button, that lets the user remove all active subscriptions and then the selected plan with one click:

![image](https://user-images.githubusercontent.com/11555574/149156886-64d2f4a1-fe99-4209-981a-eeacb417dcf8.png)

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

#### Plan without Marketplace subscriptions
* Purchase a plan
* Navigate to `/me/purchases` and click on the plan
* Click on the `Remove <plan name>` button
* Verify that the new modal is not shown 

#### Remove all subscriptions ([video](https://d.pr/v/IMTRcf))
* Purchase a plan and one or more subscriptions
* Navigate to `/me/purchases` and click on the plan
* Click on the `Remove <plan name>` button
* Verify that the modal is shown, and the correct subscriptions are listed on it
* Click on the `Remove Plan & All Subscriptions` button on the modal
* Fill out the feedback form and click on `Remove Plan` button
* Verify that all Marketplace subscriptions and the plan are removed, and you're redirected back to the `/me/purchases` page

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes #59661

gcsecsey committed:
add dialog to remove subscriptions before plan

gcsecsey committed:
no need to filter for siteId, purchases are already filtered

gcsecsey committed:
fix types and button states

sirbrillig commented on client/me/purchases/marketplace-subscriptions-dialog/index.tsx:
```diff
@@ -0,0 +1,108 @@
+			onClose={ closeDialog }
+		>
+			<Fragment>
+				<FormSectionHeading>
```
There's a type error here, which is why the unit tests are failing:

```
81:  Property 'className' is missing in type '{ children: ReactChild; }' but required in type '{ [x: string]: any; className: any; children: any; }'.
```

The problem is that the definition of `FormSectionHeading` is in JS and TS is assuming that all its props are required. You can resolve this either by adding JSDoc to that function or setting a default prop for `className` like this:

```js
const FormSectionHeading = ( { className = '', children, ...otherProps } ) => (
```

gcsecsey replied:
Great tips, thanks @sirbrillig! 🎉 I've added JSDoc annotations, now the tests pass ✅

gcsecsey committed:
add JSDoc type annotations to FormSectionHeading

cpapazoglou commented on client/me/purchases/remove-purchase/index.jsx:
```diff
@@ -312,8 +314,25 @@ class RemovePurchase extends Component {
+				isDialogVisible={ this.state.isDialogVisible }
+				isRemoving={ this.state.isRemoving }
+				closeDialog={ this.closeDialog }
+				removePlan={ this.removePurchase }
```
Instead of directly removing the plan purchase we should render the initial plan dialog so that we don't introduce regressions

https://github.com/Automattic/wp-calypso/blob/96e744b73923836a8fccda12419833bd2a1a1b3d/client/me/purchases/remove-purchase/index.jsx#L245-L259

gcsecsey committed:
always get siteId from the selected purchase

gcsecsey committed:
update dialog buttons, redirect to plan dialog

- remove `remove plan` option, add `remove plan & all subscriptions` option from dialog
- make the remove all option redirect to the original plan dialog, to be able to gather user feedback

cpapazoglou commented on client/me/purchases/marketplace-subscriptions-dialog/index.tsx:
```diff
@@ -0,0 +1,76 @@
+				</FormSectionHeading>
+				<p>
+					{ translate(
+						'When you downgrade your plan, the following subscription will remain active:',
```
```suggestion
						'The following subscriptions depend on your plan:',
```

<img width="1029" alt="SS 2022-01-12 at 20 39 17" src="https://user-images.githubusercontent.com/12430020/149201732-abcaa41c-09ae-4cde-831c-b1fbd625b6c1.png">

gcsecsey replied:
Good point, sound much more logical like this 😃

JulianBustamante reviewed (approved):
LGTM! I tested it, and it worked as expected! [Video](https://d.pr/v/uEUTmo)

gcsecsey committed:
Update client/me/purchases/marketplace-subscriptions-dialog/index.tsx

Co-authored-by: Harris Papazoglou <hrrsppzgl@gmail.com>

gcsecsey committed:
update plural translation

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7082809

Thank you @gcsecsey for including a screenshot in the description! This is really helpful for our translators.

tyxla commented on client/components/forms/form-section-heading/index.jsx:
```diff
@@ -1,7 +1,16 @@
 import classnames from 'classnames';
+import { ReactChild, ReactElement } from 'react';
```
Technically we don't use those in this component.

Could you instead use the `@param {import('react').ReactChild}` syntax instead and avoid importing them unnecessarily here?

gcsecsey replied:
Great point, thanks @tyxla! I've created [a follow-up PR to handle this](https://github.com/Automattic/wp-calypso/pull/60134), please review.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7082809).
